[Concurrent Segment Search][BUG] Incorrect slice collectors to leaves association with profile enabled

**Describe the bug**
In the concurrent search case, we found there is no `associateCollectorToLeaves()` called for `FilteredCollector:getLeafCollector()` to build the scorer. This issue caused the issue to create profile result.
```
ContextIndexSearcher.searchLeaf() collector.getLeafCollector() -> FiliteredCollectorManager.newCollector() -> FilteredCollector.getLeafCollector() -> scorerSupplier(context)
```
**To Reproduce**
```
REPRODUCE WITH: ./gradlew 'null' --tests "org.opensearch.search.query.QueryProfilePhaseTests" -Dtests.method="testIndexSortingEarlyTermination [p0=5 p1=org.opensearch.search.query.ConcurrentQueryPhaseSearcher@487db668]" -Dtests.seed=82745E9780FC847D -Dtests.security.manager=false -Dtests.locale=fr-FR -Dtests.timezone=Kwajalein -Druntime.java=18
```
```
java.lang.AssertionError: OpenSearchException[Unexpected size: 6 of leaf breakdown in ConcurrentQueryProfileBreakdown of rewritten query for a leaf.]

	at __randomizedtesting.SeedInfo.seed([82745E9780FC847D:FAAC3E093A92215F]:0)
	at org.opensearch.search.profile.query.ConcurrentQueryProfileBreakdown.toBreakdownMap(ConcurrentQueryProfileBreakdown.java:95)
	at org.opensearch.search.profile.query.ConcurrentQueryProfileTree.createProfileResult(ConcurrentQueryProfileTree.java:44)
	at org.opensearch.search.profile.query.ConcurrentQueryProfileTree.createProfileResult(ConcurrentQueryProfileTree.java:24)
	at org.opensearch.search.profile.AbstractInternalProfileTree.doGetTree(AbstractInternalProfileTree.java:183)
	at org.opensearch.search.profile.AbstractInternalProfileTree.getTree(AbstractInternalProfileTree.java:155)
	at org.opensearch.search.profile.query.ConcurrentQueryProfileTree.getTree(ConcurrentQueryProfileTree.java:77)
	at org.opensearch.search.profile.query.ConcurrentQueryProfiler.getTree(ConcurrentQueryProfiler.java:74)
	at org.opensearch.search.profile.SearchProfileShardResults.buildShardResults(SearchProfileShardResults.java:202)
	at org.opensearch.search.query.QueryProfilePhaseTests.assertProfileData(QueryProfilePhaseTests.java:1494)
	at org.opensearch.search.query.QueryProfilePhaseTests.assertProfileData(QueryProfilePhaseTests.java:1483)
	at org.opensearch.search.query.QueryProfilePhaseTests.testIndexSortingEarlyTermination(QueryProfilePhaseTests.java:857)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:577)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
	at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
	at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
	at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
	at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
	at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
	at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
	at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
	at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: OpenSearchException[Unexpected size: 6 of leaf breakdown in ConcurrentQueryProfileBreakdown of rewritten query for a leaf.]
	... 49 more
```
**Expected behavior**
A clear and concise description of what you expected to happen.

**Plugins**
Standard

**Screenshots**
N/A

**Host/Environment (please complete the following information):**
 - CI

**Additional context**
Add any other context about the problem here.

--- a/server/src/main/java/org/opensearch/common/lucene/search/FilteredCollector.java
+++ b/server/src/main/java/org/opensearch/common/lucene/search/FilteredCollector.java
@@ -40,6 +40,7 @@ import org.apache.lucene.search.ScorerSupplier;
 import org.apache.lucene.search.Weight;
 import org.apache.lucene.util.Bits;
 import org.opensearch.common.lucene.Lucene;
+import org.opensearch.search.profile.query.ProfileWeight;
 
 import java.io.IOException;
 
@@ -64,6 +65,9 @@ public class FilteredCollector implements Collector {
 
     @Override
     public LeafCollector getLeafCollector(LeafReaderContext context) throws IOException {
+        if (filter instanceof ProfileWeight) {
+            ((ProfileWeight) filter).associateCollectorToLeaves(context, collector);
+        }
         final ScorerSupplier filterScorerSupplier = filter.scorerSupplier(context);
         final LeafCollector in = collector.getLeafCollector(context);
         final Bits bits = Lucene.asSequentialAccessBits(context.reader().maxDoc(), filterScorerSupplier);

--- a/server/src/main/java/org/opensearch/search/profile/query/ConcurrentQueryProfileBreakdown.java
+++ b/server/src/main/java/org/opensearch/search/profile/query/ConcurrentQueryProfileBreakdown.java
@@ -83,7 +83,12 @@ public final class ConcurrentQueryProfileBreakdown extends ContextualProfileBrea
             // creates a new weight and breakdown map for each rewritten query. This new breakdown map captures the timing information for
             // the new rewritten query. The sliceCollectorsToLeaves is empty because this breakdown for rewritten query gets created later
             // in search leaf path which doesn't have collector. Also, this is not needed since this breakdown is per leaf and there is no
-            // concurrency involved. An empty sliceCollectorsToLeaves could also happen in the case of early termination.
+            // concurrency involved.
+            assert contexts.size() == 1 : new OpenSearchException(
+                "Unexpected size: "
+                    + contexts.size()
+                    + " of leaves breakdown in ConcurrentQueryProfileBreakdown of rewritten query for a leaf."
+            );
             AbstractProfileBreakdown<QueryTimingType> breakdown = contexts.values().iterator().next();
             queryNodeTime = breakdown.toNodeTime() + createWeightTime;
             maxSliceNodeTime = 0L;

--- a/server/src/main/java/org/opensearch/search/profile/query/ConcurrentQueryProfileBreakdown.java
+++ b/server/src/main/java/org/opensearch/search/profile/query/ConcurrentQueryProfileBreakdown.java
@@ -84,11 +84,9 @@ public final class ConcurrentQueryProfileBreakdown extends ContextualProfileBrea
             // the new rewritten query. The sliceCollectorsToLeaves is empty because this breakdown for rewritten query gets created later
             // in search leaf path which doesn't have collector. Also, this is not needed since this breakdown is per leaf and there is no
             // concurrency involved.
-            assert contexts.size() == 1 : new OpenSearchException(
-                "Unexpected size: "
-                    + contexts.size()
-                    + " of leaves breakdown in ConcurrentQueryProfileBreakdown of rewritten query for a leaf."
-            );
+            assert contexts.size() == 1 : "Unexpected size: "
+                + contexts.size()
+                + " of leaves breakdown in ConcurrentQueryProfileBreakdown of rewritten query for a leaf.";
             AbstractProfileBreakdown<QueryTimingType> breakdown = contexts.values().iterator().next();
             queryNodeTime = breakdown.toNodeTime() + createWeightTime;
             maxSliceNodeTime = 0L;

